MCR-2856 JobQueue improvements

sebhofmann opened the pull request:
* avoid naming things like slaveFactory
* reduced complexity of MCRJobManager#run method and other methods
* moved most configuration options to interface
* added MAX_TRIES status, exception and tries collumns to database
* change job status to MAX_TRIES after they run n times (configurable)
* added commands to list and reset jobs which have job status MAX_TRIES
* removed single queue mode (makes code more complex and nobody should have used it)
* added more tests
* do not reset processing jobs every x minutes, reset them once at startup
* extracted cronjob code from base to new module required adding `<mapping-file>META-INF/mycore-jobqueue-mappings.xml</mapping-file>` to persistence.xml
* moved jersey code to rest api endpoint /api/jobqueue
* added vue based gui for jobqueues


[Link to jira](https://mycore.atlassian.net/browse/MCR-2856).

sebhofmann committed:
MCR-2856 JobQueue improvements

* avoid naming things like slaveFactory
* reduced complexity of MCRJobManager#run method and other methods
* moved most configuration options to interface
* added MAX_TRIES status, exception and tries collumns to database
* change job status to MAX_TRIES after they run n times (configurable)
* added commands to list and reset jobs which have job status MAX_TRIES
* removed single queue mode (makes code more complex and nobody should have used it)
* added more tests
* do not reset processing jobs every x minutes, reset them once at startup
* extracted cronjob code from base to new module required adding <mapping-file>META-INF/mycore-jobqueue-mappings.xml</mapping-file> to persistence.xml
* moved jersey code to rest api endpoint /api/jobqueue
* added vue based gui for jobqueues

sebhofmann committed:
Fix codacy issues

sebhofmann committed:
MCR-2856 fix all javadoc warnings and added copyright notice

sebhofmann committed:
MCR-2856 Added deprecated properties

sebhofmann committed:
MCR-2856 end deprecated properties with new line

sebhofmann committed:
MCR-2856 remove deprecated properties from tests

Mewel reviewed (changes requested):
* MCRJobRunnable#run() is to long
* MCRJobRunnable#run() setStatus(MCRProcessableStatus.failed) is never set
* MCRJobThreadStarter constructor is to long -> should use static inner classes for MCRJobQueueEventListener and ThreadPoolExecutor
* MCRTestJobAction class should be moved to test -> maybe rename because there is already one in test
* MCRJobThreadStarterTest#testRun() 87 -> try to avoid waiting in test cases if possible -> this is very annoying for building mycore

minor code improvements:
* some LOGGER's and other member variables maybe final
* MCRJobQueue 226 -> return Collections.emptyIterator();
* MCRJobResetter 45 -> private final Function<Class<?>, Queue<MCRJob>> queueResolver;
* MCRJobThreadStarter 203 -> null check can be removed
* MCRJobDAOJPAImplTest 271 and multiple others -> allJobs.forEach(em::persist);
MCRJobQueueTest 48 -> MCRJobQueueEventListener listener = notifiedJobs::add;

sebhofmann committed:
MCR-2856 add improvements from review

sebhofmann commented:
> * MCRJobRunnable#run() is to long

I extracted handleJobExecutionException

> * MCRJobRunnable#run() setStatus(MCRProcessableStatus.failed) is never set

Its set by setError(Throwable method)

> * MCRJobThreadStarter constructor is to long -> should use static inner classes for MCRJobQueueEventListener and ThreadPoolExecutor

Extracted ListenerNotifier, JobThreadFactory and ActiveCountingThreadPoolExecutor to static inner class 

> * MCRTestJobAction class should be moved to test -> maybe rename because there is already one in test
> * MCRJobThreadStarterTest#testRun() 87 -> try to avoid waiting in test cases if possible -> this is very annoying for building mycore

Removed "Production" TestJobAction and related Command

> 
> minor code improvements:
> 
> * some LOGGER's and other member variables maybe final

Set all loggers to final

> * MCRJobQueue 226 -> return Collections.emptyIterator();

Done

> * MCRJobResetter 45 -> private final Function<Class<?>, Queue> queueResolver;

Changed to ? extends MCRJobAction

> * MCRJobThreadStarter 203 -> null check can be removed

Done

> * MCRJobDAOJPAImplTest 271 and multiple others -> allJobs.forEach(em::persist);

Done 

>   MCRJobQueueTest 48 -> MCRJobQueueEventListener listener = notifiedJobs::add;

Done

sebhofmann committed:
MCR-2856 improve transaction handling

sebhofmann committed:
MCR-2856 fix codacy issues

sebhofmann committed:
MCR-2856 added comment to test

sebhofmann committed:
MCR-2856 reduced wait time

sebhofmann committed:
MCR-2856 improve reset time handling

sebhofmann committed:
MCR-2856 improve javadocs

yagee-de commented on mycore-jobqueue/src/main/java/org/mycore/services/queuedjob/MCRJobRunnable.java:
```diff
@@ -0,0 +1,194 @@
+
+    private void handleException(MCRJob localJob, Exception executionException) {
+        try (StringWriter sw = new StringWriter(); PrintWriter pw = new PrintWriter(sw);) {
+            executionException.printStackTrace(pw);
```
@sonatype-lift ignore

sebhofmann committed:
MCR-2856 dont use LinkedList

sebhofmann committed:
MCR-2856 optimize imports

yagee-de reviewed (approved):
change requests from @Mewel applied and added changes for sonatype and organized imports
